[WGSL] Implement entire WGSL AST

djg opened the pull request:
#### 664eceb30ec0355555f0781181f2b312f64896e2
<pre>
[WGSL] Implement entire WGSL AST
<a href="https://bugs.webkit.org/show_bug.cgi?id=246559">https://bugs.webkit.org/show_bug.cgi?id=246559</a>
rdar://problem/101198259

Reviewed by Tadeu Zagallo.

Implement the entirety of an AST to represent WGSL programs as specified in
<a href="https://bugs.webkit.org/show_bug.cgi?id=246559.">https://bugs.webkit.org/show_bug.cgi?id=246559.</a>

* The naming convention is to use full forwards for names, not abbreviations, so
  `Structure` not `Struct`.
* All node kinds have been unified into a global `NodeKind`. `TypeCastTraits`
  are implemented for AST nodes, and `is&lt;AST::Node&gt;(node)` pattern is used
  instead of direct comparison of `node.kind()` to enum values.
* Identifier has been added. It&apos;s current a wrapper around `WTF::String` but
  could become an interned string type, such as `WTF::AtomString` in the future.
* GlobarVars are just `Variable`s. What makes them global is that they&apos;re
  accessible from the `ShaderModule`.

Differences from the AST tree outlined in the bug:
* The AST in the bug eliminated the abstract integer types. It&apos;s unclear whether
  those can be removed, so they&apos;ve been kept as distinct node types.
* Due to lack of node hierarchy representation for types, `Declaration` is kept
  as a base-class for `Function`, `Structure` and `Variable`.
* `TypeDecl` really represented the spelling of type name and has been renamed
   to `TypeName`, with the eventual goal of collapsing named, parameterized and
   array types names into one node type. Due to lack of type hierarchy,
   StructDecl and ReferenceDecl have been kept but renamed StructTypeName and
   ReferenceTypeName even though this doesn&apos;t really fit.

Canonical link: <a href="https://commits.webkit.org/259661@main">https://commits.webkit.org/259661@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d79ef050476c4a5fcf44eaadfdfc19bee2c2bbf6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105613 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14703 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38487 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114842 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109515 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16109 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/5903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/97885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111367 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12250 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/95227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/97885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/94114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/26864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/97885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/95350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7972 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/28219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/93512 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5756 "Built successfully and passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/8477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/5903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30399 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14091 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/47763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/102224 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6672 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/10028 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25485 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTTypeName.h:
```diff
@@ -26,74 +26,75 @@
 };
 
-class ArrayType final : public TypeDecl {
+class ArrayTypeName : public TypeName {
```
I agree that `TypeDecl` wasn't right, since this isn't a declaration, but I don't think type name is accurate either. i.e. `array<T, n>` isn't naming a type, it's constructing a type. Maybe we could rename the base class to be `AST::Type`, and `AST::NamedTypedName -> AST::TypeName`. WDYT?

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTAbstractFloatLiteral.h:
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
```
nit: 2023

djg replied:
Well it was originally written in 2022. :-)

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTAbstractFloatLiteral.h:
```diff
@@ -0,0 +1,55 @@
+// Literal floats without size prefix; these floats are double-precision
+// floating point numbers.
+// https://gpuweb.github.io/gpuweb/wgsl/#abstractfloat
+class AbstractFloatLiteral final : public Expression {
```
Not for this PR, but just a thought: should we unify all literals? Or at least float literals and integer literals?

djg replied:
I thought of this too. In the AST we'd store the textual representation of the literal and in the type checking passed we'd check that it was a valid numeric representation for the required type.

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTBinaryExpression.cpp:
```diff
@@ -0,0 +1,61 @@
+void printInternal(PrintStream& out, BinaryOperation op)
+{
+    constexpr ASCIILiteral binaryOperationNames[] = {
+        "+"_s,
```
I think @litherum commented about enforcing the ordering here in a PR of mine. The JSC way would be to use a macro to define all the operations, and reuse the macro to print it.

djg replied:
I had toyed with the same idea but this patch was already quite large.

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTBinaryExpression.h:
```diff
@@ -26,36 +26,62 @@
     }
 
-    Kind kind() const override;
+    NodeKind kind() const override { return NodeKind::BinaryExpression; }
```
Just curious, why did you move these inline?

djg replied:
No idea. I think I should move them out-of-line in a follow up patch.

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTCallExpression.h:
```diff
@@ -26,41 +26,38 @@
     //   * Identifier that refers to a type alias.
     //   * Identifier that refers to a function.
-    Ref<TypeDecl> m_target;
+    TypeName::Ref m_target;
```
`TypeName` should never be used to refer to a function name, right?

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTDecrementIncrementStatement.h:
```diff
@@ -0,0 +1,57 @@
+
+namespace WGSL::AST {
+
+class DecrementIncrementStatement final : public Statement {
```
Shouldn't this be called e.g. `PostfixStatement`?

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTIdentifier.h:
```diff
@@ -0,0 +1,84 @@
+    operator String&() { return m_id; }
+    operator const String&() const { return m_id; }
+
+    const String& id() const { return m_id; }
```
nit: to avoid the abbreviation, should we call this `name()`?

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTIdentifierExpression.h:
```diff
@@ -26,26 +26,27 @@
 
 namespace WGSL::AST {
 
 class IdentifierExpression final : public Expression {
```
It feels a bit weird how similar this is to `AST::Identifier` now, but maybe it's inevitable?

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTOverrideValue.h:
```diff
@@ -0,0 +1,65 @@
+
+namespace WGSL::AST {
+
+class OverrideValue : public Value {
```
I don't understand what makes this a value? This looks like an assignment, which according to the grammar I believe is a declaration?

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTParameterValue.h:
```diff
@@ -0,0 +1,71 @@
+    ArgumentBuffer,
+};
+
+class ParameterValue final : public Value {
```
Same question as before: why do these need to be grouped under `Value`? Is there any place using `Value` which accepts either an override or a `Parameter`?

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTStructure.h:
```diff
@@ -66,37 +41,37 @@ enum class StructRole : uint8_t {
 };
 
-class StructDecl final : public Decl {
+class Structure final : public Declaration {
```
Just a nit, but I think this is one of the cases according to the style guide where the abbreviation is the preferred form. Maybe @litherum has strong feelings about this?

tadeuzagallo commented on Source/WebGPU/WGSL/AST/ASTValue.h:
```diff
@@ -27,32 +27,28 @@
 namespace WGSL::AST {
 
-class Decl : public Node {
+class Value : public Node {
```
Same comment as before, seems like this node doesn't need to exist, unless I missed something.

tadeuzagallo commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -399,31 +395,60 @@ void FunctionDefinitionWriter::visit(AST::UnaryExpression& unary)
+    case AST::BinaryOperation::LessEqual:
+    case AST::BinaryOperation::ShortCircuitAnd:
+    case AST::BinaryOperation::ShortCircuitOr:
+        // FIXME: Implement these
```
nit: I usually like to `ASSERT_NOT_REACHED` so know exactly what's missing when we get to them

tadeuzagallo commented on Source/WebGPU/WGSL/SourceSpan.h:
```diff
@@ -41,25 +41,28 @@ struct SourceSpan {
     unsigned m_length;
 
-    SourceSpan(unsigned line, unsigned lineOffset, unsigned offset, unsigned length)
+    static constexpr SourceSpan empty() { return { 0, 0, 0, 0 }; }
```
Thank you 😅

tadeuzagallo reviewed (approved):
This is a pretty large diff, and I left a lot of comments, but please don't feel like you have to address all of them before landing. Specially considering the WebKit workflow, it makes it very hard to see what changed while iterating.

Overall I like the changes, I think there's some things we can still change going forward, but as I said, I'm all for landing this asap it adds a lot of things we need and to keep rebasing it is going to be a pain.

djg committed:
[WGSL] Implement entire WGSL AST
https://bugs.webkit.org/show_bug.cgi?id=246559
rdar://problem/101198259

Reviewed by Tadeu Zagallo.

Implement the entirety of an AST to represent WGSL programs as specified in
https://bugs.webkit.org/show_bug.cgi?id=246559.

* The naming convention is to use full forwards for names, not abbreviations, so
  `Structure` not `Struct`.
* All node kinds have been unified into a global `NodeKind`. `TypeCastTraits`
  are implemented for AST nodes, and `is<AST::Node>(node)` pattern is used
  instead of direct comparison of `node.kind()` to enum values.
* Identifier has been added. It's current a wrapper around `WTF::String` but
  could become an interned string type, such as `WTF::AtomString` in the future.
* GlobarVars are just `Variable`s. What makes them global is that they're
  accessible from the `ShaderModule`.

Differences from the AST tree outlined in the bug:
* The AST in the bug eliminated the abstract integer types. It's unclear whether
  those can be removed, so they've been kept as distinct node types.
* Due to lack of node hierarchy representation for types, `Declaration` is kept
  as a base-class for `Function`, `Structure` and `Variable`.
* `TypeDecl` really represented the spelling of type name and has been renamed
   to `TypeName`, with the eventual goal of collapsing named, parameterized and
   array types names into one node type. Due to lack of type hierarchy,
   StructDecl and ReferenceDecl have been kept but renamed StructTypeName and
   ReferenceTypeName even though this doesn't really fit.

Canonical link: https://commits.webkit.org/259661@main
